storage/registries.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package storage
+
+import (
+	"sort"
+
+	"github.com/juju/errors"
+)
+
+// ChainedProviderRegistry is storage provider registry that combines
+// multiple storage provider registries, chaining their results. Registries
+// earlier in the chain take precedence.
+type ChainedProviderRegistry []ProviderRegistry
+
+// StorageProviderTypes implements ProviderRegistry.
+func (r ChainedProviderRegistry) StorageProviderTypes() []ProviderType {
+	var result []ProviderType
+	for _, r := range r {
+		result = append(result, r.StorageProviderTypes()...)
+	}
+	return result
+}
+
+// StorageProvider implements ProviderRegistry.
+func (r ChainedProviderRegistry) StorageProvider(t ProviderType) (Provider, error) {
+	for _, r := range r {
+		p, err := r.StorageProvider(t)
+		if err == nil {
+			return p, nil
+		}
+		if errors.IsNotFound(err) {
+			continue
+		}
+		return nil, errors.Annotatef(err, "getting storage provider %s", t)
+	}
+	return nil, errors.NotFoundf("storage provider %s", t)
+}
+
+// StaticProviderRegistry is a storage provider registry with a statically
+// defined set of providers.
+type StaticProviderRegistry struct {
+	Providers map[ProviderType]Provider
```

Comment:
Should this be unexported and an accessor method added which copies and returns the map? Do we want external actors able to mess with the contents?
If it is to be exported, needs a comment I guess.